### PR TITLE
Pass toolchain's root into ExternalProject

### PR DESCRIPTION
**Pull request by Ignition**
### Description

With the new toolchain-v5 there was an issue where 3rd party lib was being built with incorrect libzstd on some developers machines. This causes linking issues.

Fix is to provide the toolchain's root via `CMAKE_PREFIX_PATH` in our usage of `ExternalProject_Add`.

Our toolchain environment file can be updated to introduce `MG_TOOLCHAIN_ROOT` to make this work.

@MarkoBarisic @gitbuda should we republish toolchain-v5 to have `MG_TOOLCHAIN_ROOT`?

[master < Task] PR
- [x] Provide the full content or a guide for the final git message
    - "Pass toolchain's root into ExternalProject"

### Documentation checklist
- [x] Add the documentation label tag
- [x] Add the bug / feature label tag
- [x] Add the milestone for which this feature is intended
    - If not known, set for a later milestone
- [ ] Write a release note, including added/changed clauses
    - **[Release note text]**
- [ ] Link the documentation PR here
    - **[Documentation PR link]**
- [ ] Tag someone from docs team in the comments

### CODE REVIEW

File: libs/CMakeLists.txt [hunk 1]
```diff
@@ -84,29 +79,44 @@ function(add_external_project name)
...
-    CONFIGURE_COMMAND ${KW_CONFIGURE_COMMAND}
-    CMAKE_ARGS -DCMAKE_BUILD_TYPE=Release
-    -DCMAKE_CXX_COMPILER=${CMAKE_CXX_COMPILER}
-    -DCMAKE_INSTALL_PREFIX=${source_dir}
-    ${KW_CMAKE_ARGS}
-    INSTALL_COMMAND ${KW_INSTALL_COMMAND}
-    BUILD_COMMAND ${KW_BUILD_COMMAND})
+          PREFIX ${source_dir} SOURCE_DIR ${source_dir}
+          BUILD_IN_SOURCE ${build_in_source}
+          LIST_SEPARATOR " "
```
Comment (mathstuf) [inline libs/CMakeLists.txt:84]: This is far too common to work. You should use something like `-+-` or the like that has very little chance of occurring other than as "I'm smuggling a `;` through to `ExternalProject` for `*_COMMAND` arguments".

File: libs/CMakeLists.txt [hunk 2]
```diff
@@ -257,13 +270,13 @@ import_external_library(pulsar STATIC
   -DCMAKE_INSTALL_PREFIX=${CMAKE_CURRENT_SOURCE_DIR}/pulsar/install
   -DCMAKE_CXX_COMPILER=${CMAKE_CXX_COMPILER}
   -DCMAKE_C_COMPILER=${CMAKE_C_COMPILER}
+  "-DCMAKE_PREFIX_PATH=${TOOLCHAIN_ROOT};${PROTOBUF_ROOT}"
```
Comment (mathstuf) [inline libs/CMakeLists.txt:273]: Use whatever you set for `LIST_SEPARATOR` here instead of `;`.

File: libs/CMakeLists.txt [hunk 3]
```diff
@@ -84,29 +79,44 @@ function(add_external_project name)
...
+          PREFIX ${source_dir} SOURCE_DIR ${source_dir}
+          BUILD_IN_SOURCE ${build_in_source}
+          LIST_SEPARATOR " "
+          CONFIGURE_COMMAND ${KW_CONFIGURE_COMMAND}
+          CMAKE_ARGS -DCMAKE_BUILD_TYPE=Release
+          -DCMAKE_CXX_COMPILER=${CMAKE_CXX_COMPILER}
+          -DCMAKE_INSTALL_PREFIX=${source_dir}
+          ${KW_CMAKE_ARGS}
+          INSTALL_COMMAND ${KW_INSTALL_COMMAND}
+          BUILD_COMMAND ${KW_BUILD_COMMAND})
```
Comment (mathstuf) [inline libs/CMakeLists.txt:91]: Note that by doing this, you suppress the default command generation for CMake projects. If everything is CMake, I'd recommend just not mentioning these at all.

**Review by gitbuda — APPROVED** [review]
Nice!
@mathstuf thanks a lot for your contribution!
@MarkoBarisic, yes, we should extend the `activate` script by adding `MG_TOOLCHAIN_ROOT` 👀

File: libs/CMakeLists.txt [hunk 4]
```diff
@@ -265,7 +286,7 @@ import_external_library(pulsar STATIC
   -DLINK_STATIC=ON
   -DPROTOC_PATH=${PROTOBUF_ROOT}/bin/protoc
   -DBOOST_ROOT=${BOOST_ROOT}
-  -DCMAKE_PREFIX_PATH=${PROTOBUF_ROOT}
+  "-DCMAKE_PREFIX_PATH=$<$<BOOL:MG_TOOLCHAIN_ROOT>:${MG_TOOLCHAIN_ROOT}${LIST_SEP}>${PROTOBUF_ROOT}"
```
Comment (mathstuf) [inline libs/CMakeLists.txt:289]:
```suggestion
  "-DCMAKE_PREFIX_PATH=$<$<BOOL:${MG_TOOLCHAIN_ROOT}>:${MG_TOOLCHAIN_ROOT}${LIST_SEP}>${PROTOBUF_ROOT}"
```

**Pull request merged**